group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2427,25 +2470,281 @@ private boolean hasMemberSubscriptionChanged(
+        if (!Objects.equals(oldSubscribedTopicRegex, newSubscribedTopicRegex)) {
+            log.debug("[GroupId {}] Member {} updated its subscribed regex to: {}.",
+                groupId, memberId, newSubscribedTopicRegex);
+
+            if (isNotEmpty(oldSubscribedTopicRegex) && group.numSubscribedMembers(oldSubscribedTopicRegex) == 1) {
+                // If the member was the last one subscribed to the regex, we delete the
+                // resolved regular expression.
+                records.add(GroupCoordinatorRecordHelpers.newConsumerGroupRegularExpressionTombstone(
+                    groupId,
+                    oldSubscribedTopicRegex
+                ));
+            }
 
-            if (!updatedMember.subscribedTopicRegex().equals(member.subscribedTopicRegex())) {
-                log.debug("[GroupId {}] Member {} updated its subscribed regex to: {}.",
-                    groupId, memberId, updatedMember.subscribedTopicRegex());
-                // If the regular expression has changed, we compile it to ensure that
-                // its syntax is valid.
-                if (updatedMember.subscribedTopicRegex() != null) {
+            if (isNotEmpty(newSubscribedTopicRegex)) {
+                if (group.numSubscribedMembers(newSubscribedTopicRegex) == 0) {
+                    // If the member subscribed to a new regex, we compile it to ensure its validity.
+                    // We also trigger a refresh of the regexes in order to resolve it.
                     throwIfRegularExpressionIsInvalid(updatedMember.subscribedTopicRegex());
+                    requireRefresh = true;
+                } else {
+                    // If the new regex is already resolved, we trigger a rebalance
+                    // by bumping the group epoch.
+                    bumpGroupEpoch = group.resolvedRegularExpression(newSubscribedTopicRegex).isPresent();
                 }
-                return true;
             }
         }
-        return false;
+
+        // Conditions to trigger a refresh:
+        // 0. The group is subscribed to regular expressions.
+        // 1. There is no ongoing refresh for the group.
+        // 2. The last refresh is older than 10s.
+        // 3. The group has unresolved regular expressions.
+        // 4. The metadata image has new topics.
+
+        // 0. The group is subscribed to regular expressions. We also take the one
+        //    that the current may have just introduced.
+        if (!requireRefresh && group.subscribedRegularExpressions().isEmpty()) {
+            return bumpGroupEpoch;
+        }
+
+        // 1. There is no ongoing refresh for the group.
+        String key = group.groupId() + "-regex";
+        if (executor.isScheduled(key)) {
+            return bumpGroupEpoch;
+        }
+
+        // 2. The last refresh is older than 10s. If the group does not have any regular
+        //    expressions but the current member just brought a new one, we should continue.
```

Comment:
to confirm, this is to set a limit on how often we trigger refreshes in the case the created/deleted topics rate is high?

Reply:
That's right.